### ISSUE TITLE
Rx seems to hide the true origin of a TypeError exception raised in a selector function

### ISSUE THREAD

**Issue by jcafhe** [ReactiveX/RxPY#209]
Hello,
I'm having issues with how Rx handles TypeError exception raised in a function with Rx 1.6.0, python 3.5.3 and python 3.4.5 . The code below is a minimal example that reproduces the issue:

```python
# -*- coding: utf-8 -*-
import rx


def compute(x):
#    raise Exception('Foo')
#    raise ArithmeticError('Foo')
#    raise AttributeError('Foo')
#    raise BufferError('Foo')
#    raise EOFError('Foo')
#    raise OSError('Foo')
#    raise RuntimeError('Foo')
#    raise SystemError('Foo')
    raise TypeError('Foo') # this one is not working correctly
#    raise ValueError('Foo')
    return x


(rx.Observable
   .just(1)
   .map(lambda x: compute(x))
   .subscribe(print)
 )
```

The *compute(x)* function raises a specific Exception (comment/uncomment lines). When raising TypeError, I get a TypeError printed on console, but the traceback is weird :

```python
Traceback (most recent call last):
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "C:\DEV\pyProjects\tp_QT\roaming\issues\rx_catch_exception_simplified.py", line 22, in <module>
    .subscribe(print)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observablebase.py", line 83, in subscribe
    current_thread_scheduler.schedule(set_disposable)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 50, in schedule
    return self.schedule_relative(timedelta(0), action, state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 67, in schedule_relative
    Trampoline.run(queue)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 31, in run
    item.invoke()
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\scheduleditem.py", line 17, in invoke
    ret = self.scheduler.invoke_action(self.action, self.state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\schedulerbase.py", line 14, in invoke_action
    ret = action(self, state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\returnvalue.py", line 30, in action
    observer.on_next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 19, in on_next
    self._on_next_core(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\autodetachobserver.py", line 16, in _on_next_core
    self.observer.on_next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 19, in on_next
    self._on_next_core(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\anonymousobserver.py", line 14, in _on_next_core
    self._next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\select.py", line 33, in on_next
    observer.on_error(err)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 33, in on_error
    self._on_error_core(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\autodetachobserver.py", line 23, in _on_error_core
    self.observer.on_error(exn)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 33, in on_error
    self._on_error_core(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\anonymousobserver.py", line 17, in _on_error_core
    self._error(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\basic.py", line 33, in default_error
    raise err
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\select.py", line 31, in on_next
    result = selector(value, count[0])
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\utils.py", line 42, in func_wrapped
    return func(*args, **kw)
TypeError: <lambda>() takes 1 positional argument but 2 were given 
```

So here we've got a TypeError but the trace is not pointing to the rigth place. In fact, this is a different exception with the same etype. If I raise another exception, let's say an ArithmeticError, i get the expected behaviour with a traceback pointing to the right line:

```python
Traceback (most recent call last):
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\runpy.py", line 183, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\runpy.py", line 109, in _get_module_details
    __import__(pkg_name)
  File "C:\DEV\pyProjects\tp_QT\roaming\issues\rx_catch_exception_simplified.py", line 22, in <module>
    .subscribe(print)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observablebase.py", line 83, in subscribe
    current_thread_scheduler.schedule(set_disposable)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 50, in schedule
    return self.schedule_relative(timedelta(0), action, state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 67, in schedule_relative
    Trampoline.run(queue)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\currentthreadscheduler.py", line 31, in run
    item.invoke()
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\scheduleditem.py", line 17, in invoke
    ret = self.scheduler.invoke_action(self.action, self.state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\concurrency\schedulerbase.py", line 14, in invoke_action
    ret = action(self, state)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\returnvalue.py", line 30, in action
    observer.on_next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 19, in on_next
    self._on_next_core(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\autodetachobserver.py", line 16, in _on_next_core
    self.observer.on_next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 19, in on_next
    self._on_next_core(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\anonymousobserver.py", line 14, in _on_next_core
    self._next(value)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\select.py", line 33, in on_next
    observer.on_error(err)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 33, in on_error
    self._on_error_core(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\autodetachobserver.py", line 23, in _on_error_core
    self.observer.on_error(exn)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\observerbase.py", line 33, in on_error
    self._on_error_core(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\core\anonymousobserver.py", line 17, in _on_error_core
    self._error(error)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\basic.py", line 33, in default_error
    raise err
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\linq\observable\select.py", line 31, in on_next
    result = selector(value, count[0])
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\utils.py", line 33, in func_wrapped
    ret = fn(*args, **kw)
  File "C:\DEV\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\utils.py", line 22, in func1
    return func(arg1)
  File "C:\DEV\pyProjects\tp_QT\roaming\issues\rx_catch_exception_simplified.py", line 21, in <lambda>
    .map(lambda x: compute(x))
  File "C:\DEV\pyProjects\tp_QT\roaming\issues\rx_catch_exception_simplified.py", line 7, in compute
    raise ArithmeticError('Foo')
ArithmeticError: Foo
```
I have tried with all type of exception commented in the compute function. All of these work as expected except for TypeError.

I've looked at [line 42 in rx/internal/utils.py](https://github.com/ReactiveX/RxPY/blob/63c7e4bd6cc394365d2068dfbe0ae60401beabfc/rx/internal/utils.py#L42) and I think there is an issue with the adapt_call() function. From what I understand, adapt_call() is used to 'reduce' the number of arguments that will be provided to the selector function using a try-except-TypeError check to detect how many parameters are accepted. So, there should be a conflict between a TypeError raised because of two much arguments and a TypeError raised inside the selector. With the example above :
* On the first try-except-TypeError evaluation with func1 (only 1 parameter), the selector function *lambda x: compute(x)* raises inside its body a TypeError which leads to confusion because selector has only one parameter and Rx supposes it has more than one. 
* On the 2nd try-except-TypeError  with func2 (2 parameters), a TypeError is raised since selector has only one parameter.
* finally, at line 42, selector function is feeded with too much arguments (two) since the first check was missed which leads to a 'TypeError: <lambda>() lambda takes 1 positional argument but 2 were given'.  I've checked with pdb with the example above and it confirms that args contains 2 values.

```python
TypeError: <lambda>() takes 1 positional argument but 2 were given
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> c:\dev\anaconda3_4.2.0\envs\bigboom\lib\site-packages\rx\internal\utils.py(42)func_wrapped()
-> return func(*args, **kw)
(Pdb) args
args = (1, 0)
kw = {}
(Pdb) kw
{}
```

After looking at rx/internals/utils.py history, it appears that before simplification done on commit 63c7e4bd6cc394365d2068dfbe0ae60401beabfc,  there were 2 versions of *adapt_call(selector)* function. One of these did not rely on try-exept-TypeError and used *inspect.getargspec(func)* instead. 

I would suggest to use, for python 3, the *inspect.signature(func)* and *inspect.Signature.bind(\*args, \*\*kwargs)* since *inspect.getargspec(func)* is deprecated. 

According to the [documentation](https://docs.python.org/3.5/library/inspect.html#inspect.signature), *inspect.signature(func)* return a Signature object from a function.  Then, *Signature.bind(\*args, \*\*kwargs)* raises a TypeError if args and kwargs don't match the signature.

~~Maybe all of this is related to 51.~~

**Comment by neilvyas** [comment]
I think this is related to #205 (and #213).

**Comment by mannygit** [comment]
Hi @jcafhe does head of develop branch solve the issue?

**Comment by jcafhe** [comment]
Yes it does ! I have also tried with more complex pipelines than the minimal example and it works for me.

Sorry for the delay, I am a git/github newbie. Should I close this issue now or should I wait the next release ?

**Comment by mannygit** [comment]
Thanks for the fix @neilvyas !
Good question @jcafhe . @dbrattli @thomasnield are best to answer this. It seems like we can close this out or assign it to a release milestone. It would be great to cut a minor release for this and push it to pypi :)

**Comment by markns** [comment]
I recently ran into this issue as well. Agree, it would be great to get a pypi release at some point :)

**Comment by jcafhe** [comment]
I'm closing this since this issue is fixed in the 1.6.1 release (available on PyPI).
